### DISCUSSION TITLE
Scribbles on abdominal CT data

### DISCUSSION THREAD

**Discussion by marianabb** [Project-MONAI/MONAILabel#784] [Q&A]
Hello,

I've been testing a few of the example apps in hopes of using this great tool in our environment. I've now setup a working version of both the 'radiology/segmentation_spleen' and 'radiology/segmentation' apps (our final application would be multi-label, so the 'segmentation' app is the one I'm more interested in). I am using MONAILabel version 0.4.0rc3, 3D Slicer 5.1.0 r30918.

The Scribbles module works as intended in both these applications when using the spleen task dataset. However, it doesn't generate correct results when using the abdominal CT data from [https://www.synapse.org/#!Synapse:syn3193805/wiki/217789](https://www.synapse.org/#!Synapse:syn3193805/wiki/217789). I also have the same issues when using my local abdominal CT data.

In most cases, even though foreground scribbles have been drawn, I can see the following in the logs:

`[2022-05-13 16:34:04,972] [229768] [MainThread] [INFO] (root:128) - warning: no foreground scribbles received with label 15, adding foreground scribbles to ROI centre`

This is the result after I've created a ROI and added some foreground scribbles:
![scribbles_ROI](https://user-images.githubusercontent.com/506780/168586935-4e41f479-63a3-4a35-8619-ee7606c361c5.png)

This is another result if I don't include the ROI, which maybe is more of the expected result, but still outside the frame of the original image:
![scribbles_noROI](https://user-images.githubusercontent.com/506780/168587099-4516e8a4-e563-45e7-8794-41f21d18cbbf.png)

If I add reader="ITKReader" to the LoadImaged pre-transform in class HistogramBasedGraphCut, I get a segmentation in the same frame as the original image, but still inverted (this was also generated without ROI, but if I add one the result is usually just a couple of voxels):
![scribbles_ITKReader_noROI2](https://user-images.githubusercontent.com/506780/168587179-d85f0fc0-76d2-4eb9-bd59-34f0c96e0502.png)

The Auto-segmentation module using the pretrained network works fine, so this seems to be an issue specific to the scribbles part. Any ideas on what to try next?

Thanks in advance!

**Comment by SachidanandAlle** [comment]
@masadcv can you help here..

**Comment by marianabb** [comment]
I've also tried this on the reoriented version of the abdominal CT dataset [here](https://emckclac-my.sharepoint.com/personal/k2039747_kcl_ac_uk/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fk2039747%5Fkcl%5Fac%5Fuk%2FDocuments%2FBTCV%5Freoriented&ga=1).

Using a ROI results in just a few voxels segmented in a closer, but still not correct location:
![scribbles_ROI_reoriented](https://user-images.githubusercontent.com/506780/168614583-13a3440e-a899-4d00-95a8-f5714b49ef30.png)

And without ROI, also slightly better but still inverted:
![scribbles_noROI_reoriented](https://user-images.githubusercontent.com/506780/168614856-704d28cf-b225-40ad-8767-68dc8085f7dc.png)

**Comment by diazandr3s** [comment]
Thanks for the detailed report, @marianabb 

While we wait for @masadcv to reply, have you considered using other manual annotation tools available in Slicer? Like, [Grow from seeds](https://www.youtube.com/watch?v=BJoIexIvtGo) OR the [3D brush](https://user-images.githubusercontent.com/11991079/168618528-21bec29e-e201-458d-a41b-3ff3124d6819.mp4)?

**Reply by marianabb** [reply]
Thanks for the great suggestions @diazandr3s, we will definitely also use those in our pipeline! We're a bit new to 3D Slicer so that video on how to use the 3D brush has already taught me a lot.

The main reason we'd like to use scribbles is to "smartly" edit already existing segmentations generated from our model, although maybe some of the already existing 3D Slicer segmentation tools like scissors or margin might do the job.

**Comment by diazandr3s** [comment]
Glad to hear this helped :)

**Comment by masadcv** [comment]
@marianabb many thanks for reporting this. I am looking into the issue and have successfully recreated the error using the Abdominal dataset from https://www.synapse.org/#!Synapse:syn3193805/wiki/217789

The issue seems to be  in header information for the images mentioned here. Using the default nibabel to load these images raises the following error message:
`[2022-05-16 21:26:56,001] [225596] [MainThread] [INFO] (nibabel.global:274) - pixdim[0] (qfac) should be 1 (default) or -1; setting qfac to 1`

I looked at header for one of the images in your screenshots (img0034.nii.gz) using `ITKSnap->Tools->Image Information->Metadata` (http://www.itksnap.org/) - and I can see pixdim[0] = 0 which does not make sense to me (maybe I am missing something here). Here is this information from ITKSnap:
![Screenshot from 2022-05-16 22-07-54](https://user-images.githubusercontent.com/3410225/168682842-c0e177bc-4d1d-4f36-a7bf-211a9d323d52.png)


As a comparison, here is the same information from a volume in MSD Spleen dataset (which has more plausible pxidim[0]):
![Screenshot from 2022-05-16 22-02-17](https://user-images.githubusercontent.com/3410225/168682861-bcd59e70-2900-4a99-9309-91c134e45c00.png)

I will discuss this issue with @SachidanandAlle and @diazandr3s tomorrow and try to work out a solution....

**Reply by diazandr3s** [reply]
Nifti header strikes again. As @masadcv says, this is related to Nifti headers of the BTCV dataset. 

I'm wondering how the [RasToIjkMatrix](https://github.com/Project-MONAI/MONAILabel/blob/main/plugins/slicer/MONAILabel/MONAILabel.py#L1904) is being affected by the Nifti headers?  

Perhaps a sanity check of the dataset before using it on the MONAI Label platform should be considered?

CC @lassoan @SachidanandAlle

**Reply by marianabb** [reply]
Thanks for looking into this @masadcv!

Was it enough to change the header to pixdim = 1 on the images from the Synapse dataset?

I'm getting the same problem on a .nrrd image which I believe doesn't have 'pixdim' at all. I can't share the actual image since it's from my local dataset, but here is the ITKSnap inspector on it:
![info_nrrd](https://user-images.githubusercontent.com/506780/168837596-c7ef202f-fb36-49c2-8f0a-d6baa6c2e88c.png)

**Reply by marianabb** [reply]
We've done a bit more testing with different datasets and it seems like the tool works as intended only when the original image orientation is RAS ('orientation'=[-1, 0, 0, 0, -1, 0, 0, 0,1]) . Other orientations result in similar errors in the segmentation.

I have noticed that one of the pre-transforms in class HistogramBasedGraphCut is Orientation to RAS, but maybe there are some more orientation-related changes going on deeper in the code?

**Reply by masadcv** [reply]
Many thanks for testing this! This is quite useful information.
You are absolutely right, orientation transform could be an issue here. Since the method in `HistogramBasedGraphCut` is not dependent on orientation, we can remove that orientation transform....can you try removing orientation transform in `HistogramBasedGraphCut`? This will particularly solve the flipping of segmentation issue you are experiencing (e.g. in the following image from your comment):
![https://user-images.githubusercontent.com/506780/168587179-d85f0fc0-76d2-4eb9-bd59-34f0c96e0502.png](https://user-images.githubusercontent.com/506780/168587179-d85f0fc0-76d2-4eb9-bd59-34f0c96e0502.png)

I already have this removed in a PR (https://github.com/Project-MONAI/MONAILabel/pull/763) that contains this change, but can take a short while to have this in main. This PR also adds a new scribble infer task `GMMBasedGraphCut` which has a better likelihood generation method that may be of interest to you.

Please let me know if you still experience issues with this, will be happy to solve these.

**Reply by marianabb** [reply]
Thanks for the suggestion @masadcv, I did try removing the orientation transform but it does not seem to change the results. I also tried moving the transform to before 'AddBackgroundScribblesFromROId' but it also didn't change the final output.

**Reply by masadcv** [reply]
Is it possible to share the image that has this issue, so I can recreate and have a closer look on my end?

**Reply by marianabb** [reply]
I believe any image from the Synapse dataset would reproduce the issue (I've just been clicking on the 'Next sample' button with random strategy). The last one I've been testing on is img0005.nii.gz from this dataset.

**Reply by masadcv** [reply]
One additional thing to try is loading with ITKReader (maybe you already are doing this) with orientation transform removed.

`LoadImaged(keys=["image", "label"], reader="ITKReader"),`

**Reply by masadcv** [reply]
Trying image img0005.nii.gz:
![image](https://user-images.githubusercontent.com/3410225/168878063-7477b15b-c92c-488b-88ec-cc302f761338.png)

Pre-transforms:
```
    def pre_transforms(self, data):
        return [
            LoadImaged(keys=["image", "label"], reader="ITKReader"),
            EnsureChannelFirstd(keys=["image", "label"]),
            AddBackgroundScribblesFromROId(
                scribbles="label",
                scribbles_bg_label=self.scribbles_bg_label,
                scribbles_fg_label=self.scribbles_fg_label,
            ),
            # at the moment optimisers are bottleneck taking a long time,
            # therefore scaling non-isotropic with big spacing
            Spacingd(keys=["image", "label"], pixdim=self.pix_dim, mode=["bilinear", "nearest"]),
            # Orientationd(keys=["image", "label"], axcodes="RAS"),
            ScaleIntensityRanged(
                keys="image",
                a_min=self.intensity_range[0],
                a_max=self.intensity_range[1],
                b_min=self.intensity_range[2],
                b_max=self.intensity_range[3],
                clip=self.intensity_range[4],
            ),
            MakeLikelihoodFromScribblesHistogramd(
                image="image",
                scribbles="label",
                post_proc_label="prob",
                scribbles_bg_label=self.scribbles_bg_label,
                scribbles_fg_label=self.scribbles_fg_label,
                normalise=True,
            ),
        ]

```

Note: ROI will still not work with these images as the header is wrong and affects the ROI transforms on Slicer plugin side...

**Reply by masadcv** [reply]
@diazandr3s maybe we can help @marianabb with reorienting images, so this can be resolved? Is there a code to correct header for these problematic images?

**Reply by marianabb** [reply]
I had not tried using the ITKReader again, you're right, now this is pretty close to working properly! 

I can reorient all my images to RAS for now and hopefully this will be fully resolved later :)

Thanks for all the help!

**Reply by diazandr3s** [reply]
I think it'll be great to have these images in the NRRD format instead of Nifti. There is now an NRRD reader in MONAI https://github.com/Project-MONAI/MONAI/pull/4259 and a writer in MONAI Label https://github.com/Project-MONAI/MONAILabel/pull/779

@marianabb would you like to give it a try if Nifti files don't work?

**Reply by marianabb** [reply]
I believe the root problem (and please correct me if I'm wrong) is not the image formats, but the image orientation. Without the proposed changes the tool only works properly in images with RAS orientation (Nifti or NRRD). With the proposed changes the tool works properly in both formats if no ROI is used to limit the segmentation, the ROI part is still only working properly with images in RAS orientation.

**Reply by diazandr3s** [reply]
I might be wrong, but I think the issue is the image header. 

I meant we first load the images, correct the headers and then save the images in NRRD or Nifti format
